### PR TITLE
Update ping test to not crash on failing pings (BugFix)

### PR DESCRIPTION
**Pull request by Hook25**
## Description

If the ping test fails, it will produce a traceback, this is wrong. This PR changes that leading to a correct failure path

## Resolved issues

Fixes: https://github.com/canonical/checkbox/issues/494

## Documentation

N/A

## Tests

Tests added to a new file with 100% coverage of relevant code

### CODE REVIEW

File: providers/base/bin/gateway_ping_test.py [hunk 1]
```diff
@@ -219,85 +219,125 @@ def ping(host, interface, count, deadline, verbose=False):
...
+    args = parse_args(argv)
+
     # If given host is not pingable, override with something pingable.
     host = get_host_to_ping(
-        interface=args.interface, verbose=args.verbose, default=args.host)
+        interface=args.interface, verbose=args.verbose, default=args.host
+    )
     if args.verbose:
         print(_("Checking connectivity to {0}").format(host))
-    ping_summary = None
```
Comment (fernando79513) [inline providers/base/bin/gateway_ping_test.py:280]: Would't it be better to have something like:
```
    if host:
        ping_summary = ping(
            host, args.interface, args.count, args.deadline, args.verbose
        )
    else:
        ping_summary = {"received": 0, "cause": "Could not find a host"}

    if ping_summary["received"] == 0:
        print(_("No Internet connection"))
        if ping_summary.get("cause"):
            print("Possible cause: {}".format(ping_summary["cause"]))
        return 1
```

File: providers/base/tests/test_gateway_ping_test.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    @patch("gateway_ping_test.get_host_to_ping")
+    @patch("gateway_ping_test.ping")
+    def test_invalid_arguments_count_deadline(
+        self, mock_ping, mock_get_host_to_ping
+    ):
+        with self.assertRaises(SystemExit):
+            main(["-c", "10", "-d", "8"])
+
+
```
Comment (fernando79513) [inline providers/base/tests/test_gateway_ping_test.py:82]: Flake 8
```suggestion

```

File: providers/base/tests/test_gateway_ping_test.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
+import unittest
+from unittest.mock import patch, MagicMock
```
Comment (fernando79513) [inline providers/base/tests/test_gateway_ping_test.py:2]: Unused
```suggestion
from unittest.mock import patch
```

**Review by fernando79513 — COMMENTED** [review]
Nice job with the fix and with the tests. Just some small comments.

File: providers/base/bin/gateway_ping_test.py (hunk 1 above)
Comment (Hook25) [reply providers/base/bin/gateway_ping_test.py:280]: good idea

File: providers/base/tests/test_gateway_ping_test.py (hunk 2 above)
Comment (Hook25) [reply providers/base/tests/test_gateway_ping_test.py]: we use black but sure, fogot to run it

**Review by fernando79513 — APPROVED** [review]
LGTM +1

**Pull request merged**